pkg/util/ringq/dyn_chan.go

```diff
@@ -0,0 +1,177 @@
+// It also provides a ChanStatsReader to allow querying basic stats. The
+// internal ring buffer is initially empty, and its capacity grows as more items
+// need to be stored if the receiving side is lagging behind. When the receiving
+// side catches up, the buffer will shrink to reclaim unused memory. The
+// argument minBufLen must be positive, and allows customizing the minimum
+// allocation for the internal buffer. This is useful to reduce the overhead of
+// resizing the buffer within an acceptable, permanent allocation of minBufLen
+// items ready to be used. If maxBufLen is positive, then it must be greater
+// than minBufLen, and will be used to prevent buffering more than that number
+// of items. If more items are received and the buffer already has maxBufLen,
+// then the oldest item (the one that receiving would first get) will be
+// overwritten with the new value.
+func DynChanMax[T any](minBufLen, maxBufLen int) (chan<- T, <-chan T, ChanStatsReader) {
+	if minBufLen < 1 || (maxBufLen > 0 && maxBufLen < minBufLen) {
+		panic("maxBufLen > 0 && maxBufLen < minBufLen")
+	}
+
+	var stats ChanStats
+	in := make(chan T)
+	out := make(chan T)
+	statsChan := make(chan ChanStats)
+	sr := ChanStatsReader{
+		c: statsChan,
+	}
+	if maxBufLen > 0 {
+		stats.MaxCap = maxBufLen
+	}
+
+	go func() {
+		defer close(out)
+		defer close(statsChan)
+
+		var q RingQ[T]
+		q.Max = maxBufLen
+
+		// the loop condition is that we either have items to dequeue
+		// (q.Len()>0) or that we have the possibility to receive new items to
+		// be queued (in != nil)
+		for stats.Len > 0 || in != nil {
+			// if we don't have anything in the queue, then make the dequeueing
+			// branch of the select block indefinitely by providing a nil
+			// channel, leaving only the queueing branch available
+			dequeueChan := out
+			if stats.Len == 0 {
+				dequeueChan = nil
+			}
+
+			select {
+			case v, ok := <-in: // blocks until something is queued
+				if !ok {
+					// in was closed, so if we leave it like that the next
+					// iteration will keep receiving zero values with ok=false
+					// without any blocking. So we set in to nil, so that
+					// the next iteration the select will block indefinitely on
+					// this branch of the select and leave only the dequeing
+					// branch active until all items have been dequeued
+					in = nil
+				} else {
+					if stats.Cap == 0 {
+						// initialize to the target buffer length
+						q.Grow(minBufLen)
+						stats.Cap = q.Cap()
+						stats.Allocs++
+					}
+
+					q.Enqueue(v)
+					stats.Enqueued++
+
+					l := q.Len()
+					if l == stats.Len {
+						// if after enqueueing the length remains the same, it
+						// means we overwrote an item because we overflowed the
+						// the queue
+						stats.Dropped++
+					}
+					stats.Len = l
+
+					if c := q.Cap(); stats.Cap != c {
+						stats.Cap = c
+						stats.Allocs++
+					}
+				}
+
+			case dequeueChan <- q.Peek(): // blocks if nothing to dequeue
```

Comment:
So if this was `case dequeueChan <- q.Dequeue()`, Dequeue() would still be called even if this case did not get chosen by the select statement?

Reply:
Exactly, so that would mean we have dequeued the item, meaning reading and removing it, and if the enqueue branch was picked that item would have been lost forever.

Reply:
So basically, whatever you put between the `case` and the `:` is effectively evaluated when you reach the `select` statement, and the final expression must resolve to a channel send or receive operation. You can put whatever you want and it will be evaluated, even a closure:
```go
var x int
select {
default: // this can also be in any order

// evaluate the closure and execute it, and evaluate (chan struct{})(nil)
case (chan struct{})(nil) <- func() struct{} {
	x++
	return struct{}{}
}():
}
fmt.Println(x) // prints "1"
```
https://go.dev/play/p/OCX6M6z2KTw

Reply:
Neat, I did not know that. Thanks!